File: packages/sdkwork-utils-go/i18n.go
package utils

import (
	"fmt"
	"strings"
	"time"
)

type localeSeparators struct {
	decimal  string
	grouping string
}

func separators(locale string) localeSeparators {
	if strings.EqualFold(locale, "de-DE") {
		return localeSeparators{decimal: ",", grouping: "."}
	}
	return localeSeparators{decimal: ".", grouping: ","}
}

func FormatNumberLocale(value float64, locale string, decimals int) string {
	rounded := Round(value, decimals)
	negative := rounded < 0
	absolute := Round(absFloat(rounded), decimals)
	text := fmt.Sprintf("%.*f", decimals, absolute)
	parts := strings.Split(text, ".")
	integer := parts[0]
	fraction := ""
	if len(parts) > 1 {
		fraction = parts[1]
	}
	sep := separators(locale)
	grouped := insertGrouping(integer, sep.grouping)
	if decimals > 0 {
		grouped = grouped + sep.decimal + fraction
	}
	if negative {
		return "-" + grouped
	}
	return grouped
}

func FormatDatetimeLocale(value time.Time, locale string) string {
	utc := value.UTC()
	normalized := strings.ToLower(locale)
	if strings.HasPrefix(normalized, "de") {
		return utc.Format("02.01.2006 15:04")
	}
	if strings.HasPrefix(normalized, "zh") {
		return utc.Format("2006-01-02 15:04")
	}
	return utc.Format("01/02/2006 15:04")
}

func FormatDatetimeLocaleStr(value string, locale string) (string, bool) {
	parsed, ok := ParseDatetime(value, DefaultPattern)
	if !ok {
		return "", false
	}
	return FormatDatetimeLocale(parsed, locale), true
}

func ParseNumberLocale(input string, locale string) (float64, bool) {
	trimmed := Trim(input)
	if trimmed == "" {
		return 0, false
	}
	sep := separators(locale)
	normalized := strings.ReplaceAll(strings.ReplaceAll(trimmed, sep.grouping, ""), sep.decimal, ".")
	parsed, ok := ParseNumber(normalized)
	return parsed, ok
}

func insertGrouping(integer string, grouping string) string {
	if len(integer) <= 3 {
		return integer
	}
	var parts []string
	for len(integer) > 3 {
		parts = append([]string{integer[len(integer)-3:]}, parts...)
		integer = integer[:len(integer)-3]
	}
	parts = append([]string{integer}, parts...)
	return strings.Join(parts, grouping)
}

func absFloat(value float64) float64 {
	if value < 0 {
		return -value
	}
	return value
}

File: packages/sdkwork-utils-rust/src/i18n.rs
use chrono::{DateTime, Utc};

use crate::datetime::parse_datetime;
use crate::number::round;

struct LocaleSeparators {
    decimal: char,
    grouping: char,
}

fn separators(locale: &str) -> LocaleSeparators {
    if locale.eq_ignore_ascii_case("de-DE") {
        LocaleSeparators {
            decimal: ',',
            grouping: '.',
        }
    } else {
        LocaleSeparators {
            decimal: '.',
            grouping: ',',
        }
    }
}

pub fn format_number_locale(value: f64, locale: &str, decimals: u32) -> String {
    let rounded = round(value, Some(decimals));
    let negative = rounded < 0.0;
    let absolute = round(rounded.abs(), Some(decimals));
    let decimals_usize = decimals as usize;
    let text = format!("{absolute:.decimals_usize$}");
    let parts: Vec<&str> = text.split('.').collect();
    let integer = parts[0];
    let fraction = parts.get(1).copied().unwrap_or("");
    let sep = separators(locale);

    let mut grouped = String::new();
    for (index, ch) in integer.chars().rev().enumerate() {
        if index > 0 && index % 3 == 0 {
            grouped.push(sep.grouping);
        }
        grouped.push(ch);
    }
    let grouped: String = grouped.chars().rev().collect();
    let mut result = grouped;
    if decimals > 0 {
        result.push(sep.decimal);
        result.push_str(fraction);
    }
    if negative {
        format!("-{result}")
    } else {
        result
    }
}

pub fn format_datetime_locale(value: DateTime<Utc>, locale: &str) -> String {
    let locale = locale.to_ascii_lowercase();
    if locale.starts_with("de") {
        value.format("%d.%m.%Y %H:%M").to_string()
    } else if locale.starts_with("zh") {
        value.format("%Y-%m-%d %H:%M").to_string()
    } else {
        value.format("%m/%d/%Y %H:%M").to_string()
    }
}

pub fn format_datetime_locale_str(value: &str, locale: &str) -> Option<String> {
    parse_datetime(value, None).map(|instant| format_datetime_locale(instant, locale))
}

pub fn parse_number_locale(input: &str, locale: &str) -> Option<f64> {
    let trimmed = input.trim();
    if trimmed.is_empty() {
        return None;
    }
    let sep = separators(locale);
    let normalized = trimmed.replace(sep.grouping, "").replace(sep.decimal, ".");
    normalized.parse::<f64>().ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn i18n_helpers() {
        assert_eq!(format_number_locale(1234.5, "en-US", 2), "1,234.50");
        assert_eq!(format_number_locale(1234.5, "de-DE", 2), "1.234,50");
        assert_eq!(parse_number_locale("1,234.50", "en-US"), Some(1234.5));
        assert_eq!(parse_number_locale("1.234,50", "de-DE"), Some(1234.5));
        let formatted = format_datetime_locale_str("2024-06-15T14:30:00.000Z", "en-US").unwrap();
        assert!(formatted.contains("2024"));
    }
}
